File: Subsets.js
function subsets(nums, index = 0, bag = [], res = []) {

    if(index == nums.length) {
        res.push([...bag]);
        return res;
    }
    
    bag.push(nums[index]);
    index = index+1;
    subsets(nums, index, bag, res);
    bag.pop();
    return subsets(nums, index, bag, res);
}

File: Subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {

        List<List<Integer>> res = new ArrayList<>();
        return subsetHelper(nums, 0, new ArrayList<Integer>(),  res);
    }

    private List<List<Integer>> subsetHelper(int[] nums, int index, List<Integer> bag, List<List<Integer>> res) {
        if(index == nums.length) {
          System.out.println(bag.toString());
          res.add(new ArrayList<>(bag));
          return res;
        }
        bag.add(nums[index]);
        subsetHelper(nums, index+1, bag, res);
        bag.remove(bag.size()-1);
        subsetHelper(nums, index+1, bag, res);
        return res;
    }
}
